### ISSUE TITLE
Template syntax must be validated when defined (not only when used)

### ISSUE THREAD

**Issue by kwatsen** [netmod-wg/template-reqs#4]
# Requirements

This requirement validates templates at the time they are defined, as opposed to only when they are used.   Since templates must be defined prior to being referenced, this means that unreferenced templates are validated.

# Non-Requirements

Implicit in this requirement is that templates would have to be model-specific (i.e., their YANG known) in order to be validated.  Requirements for **how** the data-model for a template would be known are NOT specified here, leaving those requirements to be specified elsewhere.

**Comment by jsterne** [comment]
Should we differentiate here between basic syntax validation (e.g. right type & range) vs semantic validation (e.g. mandatory and min-element statements, must statements, when statements, etc)?

**Comment by RobertPeschi** [comment]
>The <intended> datastore does not contain the template themselves https://github.com/netmod-wg/template-reqs/issues/15

As commented in #15, I am still struggling on why the intended data store should not contain the templates themselves

**Comment by kwatsen** [comment]
> As commented in https://github.com/netmod-wg/template-reqs/issues/15, I am still struggling on why the intended data store should not contain the templates themselves

Please keep comments with each issue/requirement.   Just because an issue/requirement is linked, doesn't mean that all its comments need to be carried over.

**Comment by robertjwills** [comment]
Expanding on Jason's point, perhaps we should distinguish between two different cases of validation failure:
* Any possible instantiation of the template would result in an expanded config that doesn't validate. In this case, the template is not allowed. This would cover basic type and range checks.
* Some instantiations of the template would be valid, others invalid. In this case, the template is allowed.

**Comment by jsterne** [comment]
I don't think we necessarily need to *preclude* a server from doing template validation (especially syntax) but I'd rather not *require* validation (or at least not anything beyond basic syntax).
